_posts/2015-12-09-docker-for-development.markdown

```diff
@@ -0,0 +1,214 @@
+
+For our applications (including Gravity) which have been dockerized, getting set up consists of a one-time install of [Docker Toolbox](https://www.docker.com/docker-toolbox) followed by running 
+
+```bash
+docker-compose build && docker-compose up
+```
+
+in the root directory of the checked-out Gravity repo. Here is a simplified version of our [docker-compose](https://docs.docker.com/compose/) setup. Because we run a web server and a delayed_job process, `docker-compose.yml` uses a `common.yml` file for shared setup:
+
+```yaml
+gravity:
+  build: .
+  environment:
+    MEMCACHE_SERVERS: memcached
+    SOLR_URL: http://solr4:8983/solr/gravity
+    MONGO_HOST: mongodb
+    ELASTICSEARCH_URL: elasticsearch:9200
+    SMTP_PORT: 1025
+    SMTP_ADDRESS: mailcatcher
+  env_file: .env
+```
+The `.env` file is used for secrets such as Amazon Web Services credentials we don't want to put into the git repository.
+
+Our `docker-compose.yml` looks like this:
+
+
+```yaml
+mongodb:
+  image: mongo:2.4
+  command: bash -c "rm -f /data/db/mongod.lock; mongod --smallfiles --quiet --logpath=/dev/null"
+  ports: 
+  - "27017:27017"
+
+solr4:
+  image: artsy/solr4
+
+memcached:
+  image: memcached
+
+elasticsearch:
+  image: artsy/elasticsearch
+  ports:
+  - "9200:9200"
+  - "9300:9300"
+
+web:
+  extends:
+    file: common.yml
+    service: gravity
+  command: script/rails s -b 0.0.0.0 -p 80
+  ports:
+  - "80:80"
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+  - mailcatcher
+
+dj:
+  extends:
+    file: common.yml
+    service: gravity
+  command: bundle exec rake jobs:work
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+
+mailcatcher:
+  image: zolweb/docker-mailcatcher
+  ports:
+  - "1080:1080"
+```
+
+The command for the MongoDB section removes a lock file that can remain in place sometimes when the container is killed. Do not use that in production! We mount the local directory into the container with a `volumes:` command, so that local changes are reloaded in the running containers.
+
+Recently, [Ashkan Nasseri](https://github.com/ashkan18) began to move our delayed jobs from [delayed_job_mongoid](https://github.com/collectiveidea/delayed_job_mongoid) to [sidekiq](http://sidekiq.org/), which brings in Redis and another process that needs to run during development. Since we are using Docker, all we have to do is add a couple of new sections to our `docker-compose.yml` file:
```

Comment:
another good example of docker benefits was our mailcatcher case, it made that setup really easy.